File: app/src/main/java/com/example/saka/myapplication/NewPostActivity.java
package com.example.saka.myapplication;

import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.AppCompatActivity;
import android.text.TextUtils;
import android.view.View;
import android.widget.EditText;
import android.widget.Toast;

import com.example.saka.myapplication.model.Post;
import com.example.saka.myapplication.model.myFirebase;

import java.util.HashMap;
import java.util.Map;

public class NewPostActivity extends AppCompatActivity {

    private EditText edtTitle;
    private EditText edtBody;
    private FloatingActionButton fab;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_new_post);

        edtTitle = (EditText) findViewById(R.id.np_edt_title);
        edtBody = (EditText) findViewById(R.id.np_edt_body);
        fab = (FloatingActionButton) findViewById(R.id.np_fab);

        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                submitPost();
                finish();
            }
        });
    }

    private void submitPost(){
        if (validatePost()) {
            // Disable button so there are no multi-post
            setEditingEnabled(false);
            Toast.makeText(getApplicationContext(), "Posting...", Toast.LENGTH_SHORT).show();

            // Start write new post
            final String uid = myFirebase.getMyUid();
            final String username = myFirebase.getUser(myFirebase.getMyUid()).getName();
            final String title = edtTitle.getText().toString();
            final String body = edtBody.getText().toString();
            writeNewPost(uid, username, title, body);
        }
    }

    private void writeNewPost(String uid, String author, String title, String body) {
        // Create new post at /user-posts/$userId/$postId and at
        // /posts/$postId simultaneously
        String key = myFirebase.getNewPushKey();
        Post post = new Post(uid, author, title, body);
        Map<String, Object> newPost = post.toMap();
        Map<String, Object> childUpdates = new HashMap<>();
        childUpdates.put("/POSTS/" + key, newPost);
        childUpdates.put("/USER-POSTS/" + uid + "/" + key, newPost);
        myFirebase.updateMyChildren(childUpdates);
    }

    private void setEditingEnabled(boolean enabled){
        edtTitle.setEnabled(enabled);
        edtBody.setEnabled(enabled);
        if (enabled){
            fab.setVisibility(View.VISIBLE);
        } else {
            fab.setVisibility(View.GONE);
        }
    }

    private boolean validatePost() {
        final String title = edtTitle.getText().toString();
        final String body = edtBody.getText().toString();
        if (TextUtils.isEmpty(title)) {
            edtTitle.setError("Required");
            return false;
        }
        if (TextUtils.isEmpty(body)) {
            edtBody.setError("Required");
            return false;
        }
        return true;
    }

}

File: app/src/main/java/com/example/saka/myapplication/ui/NewPostActivity.kt
package com.example.saka.myapplication.ui

import android.os.Bundle
import android.text.TextUtils
import android.view.View
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.example.saka.myapplication.R
import com.example.saka.myapplication.model.MyFirebase.getUser
import com.example.saka.myapplication.model.MyFirebase.myUid
import com.example.saka.myapplication.model.MyFirebase.newPushKey
import com.example.saka.myapplication.model.MyFirebase.updateMyChildren
import com.example.saka.myapplication.model.Post
import com.google.android.material.floatingactionbutton.FloatingActionButton

class NewPostActivity : AppCompatActivity() {
    private var edtTitle: EditText? = null
    private var edtBody: EditText? = null
    private var fab: FloatingActionButton? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_new_post)
        edtTitle = findViewById<View>(R.id.np_edt_title) as EditText
        edtBody = findViewById<View>(R.id.np_edt_body) as EditText
        fab = findViewById<View>(R.id.np_fab) as FloatingActionButton
        fab!!.setOnClickListener {
            submitPost()
            finish()
        }
    }

    private fun submitPost() {
        if (validatePost()) {
            // Disable button so there are no multi-post
            setEditingEnabled(false)
            Toast.makeText(applicationContext, "Posting...", Toast.LENGTH_SHORT).show()

            // Start write new post
            val uid = myUid
            val username = getUser(myUid).name ?: "name"
            val title = edtTitle!!.text.toString()
            val body = edtBody!!.text.toString()
            writeNewPost(uid, username, title, body)
        }
    }

    private fun writeNewPost(uid: String?, author: String?, title: String, body: String) {
        // Create new post at /user-posts/$userId/$postId and at
        // /posts/$postId simultaneously
        val key = newPushKey
        val post = Post(uid, author, title, body)
        val newPost = post.toMap()
        val childUpdates: MutableMap<String?, Any?> = HashMap()
        childUpdates["/POSTS/$key"] = newPost
        childUpdates["/USER-POSTS/$uid/$key"] = newPost
        updateMyChildren(childUpdates)
    }

    private fun setEditingEnabled(enabled: Boolean) {
        edtTitle!!.isEnabled = enabled
        edtBody!!.isEnabled = enabled
        if (enabled) {
            fab!!.visibility = View.VISIBLE
        } else {
            fab!!.visibility = View.GONE
        }
    }

    private fun validatePost(): Boolean {
        val title = edtTitle!!.text.toString()
        val body = edtBody!!.text.toString()
        if (TextUtils.isEmpty(title)) {
            edtTitle!!.error = "Required"
            return false
        }
        if (TextUtils.isEmpty(body)) {
            edtBody!!.error = "Required"
            return false
        }
        return true
    }
}
